[test] java tests for archive packaging

andyb-elastic opened the pull request:
Ports the first couple tests for archive distributions from the old bats
project to the new java project that includes windows platforms,
consolidating them into one test method that tests that the
distributions can be extracted and their contents verified. Includes the
zip distributions which were not tested in the bats project.

For #26741

This takes into account @rjernst's changes in #30596

andyb-elastic committed:
[test] java tests for archive packaging

Ports the first couple tests for archive distributions from the old bats
project to the new java project that includes windows platforms,
consolidating them into one test method that tests that the
distributions can be extracted and their contents verified. Includes the
zip distributions which were not tested in the bats project.

For #26741

elasticmachine commented:
Pinging @elastic/es-core-infra

andyb-elastic commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -0,0 +1,65 @@
+ * override {@link ArchiveTestCase#distribution()}. These tests should be the same across all archive distributions
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public abstract class ArchiveTestCase {
```
Originally I used [parameterized tests](https://github.com/junit-team/junit4/wiki/parameterized-tests) to serve this purpose but it seemed like this was a more easily understandable way to achieve the same thing

andyb-elastic commented on Vagrantfile:
```diff
@@ -237,6 +237,7 @@ def linux_common(config,
 
   config.vm.provision 'markerfile', type: 'shell', inline: <<-SHELL
     touch /etc/is_vagrant_vm
+    touch /is_vagrant_vm # for consistency between linux and windows
```
I'll make a follow up changing the bats tests to expect it in this location

nik9000 commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/util/Archives.java:
```diff
@@ -0,0 +1,239 @@
+        assertThat("distribution file must exist", Files.exists(distributionFile), is(true));
+        assertThat("elasticsearch must not already be installed", lsGlob(baseInstallPath, "elasticsearch*"), empty());
+
+        if (distribution.packaging == Distribution.Packaging.TAR) {
```
I wonder if this kind of thing should be moved to `ArchiveTestCase` and you subclass it for each time.

nik9000 replied:
Like you move the entire `installArchive` method to the abstract superclass.

nik9000 replied:
I think it is fine to do it as you are doing it. We could change it to my way later on if we figure it is better. I don't think it is a blocker.

I just look at your way and think "this made a ton of sense in bash scripts but I want to OO all the things in Java because I'm a Java developer." I admit that is a bad habit some times. I don't think it'd be too weird to keep the install logic in the superclass but you certainly are right that there aren't that many variants.

andyb-elastic replied:
>this made a ton of sense in bash scripts but I want to OO all the things in Java because I'm a Java developer

Yeah this was exactly what I did at first and then walked it back to here since it seemed simpler :)

I'll take a look at the other places in Archives' methods where this kind of stuff happens and see if I can find a happy medium

andyb-elastic replied:
I tried a couple variations of what we discussed. I still feel like the benefit of being able to answer the question "how are the archive distros installed" by looking in one place outweighs that place being a little gross - not in general, but in this instance because we don't expect to have many, if any additional cases. 

I agree that if it does change it would be better to do something like your suggestion. I'll also keep thinking about other ways to do it as I revise this

nik9000 commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/util/Archives.java:
```diff
@@ -0,0 +1,239 @@
+            "NOTICE.txt",
+            "LICENSE.txt",
+            "README.textile"
+        ).forEach(doc -> assertThat(es.home.resolve(doc), file(File, owner, owner, p644)));
```
This isn't passing for me:
```
Expected: "file/directory: "<File> with owner "elasticsearch" with group "elasticsearch" with posix permissions [<OWNER_READ>,<OWNER_WRITE>,<GROUP_READ>,<OTHERS_READ>]
     but: path </tmp/elasticsearch/NOTICE.txt>Has permissions [GROUP_WRITE, OWNER_READ, GROUP_READ, OWNER_WRITE, OTHERS_READ]
```

andyb-elastic replied:
Weird, it passes locally for me and in CI. Which box and test?

andyb-elastic replied:
Actually if you got a full reproduce line that would be best - I'm pretty sure the randomization doesn't affect anything about the current version installs, but best to check anyway

nik9000 replied:
I ran:
```
./gradlew -p qa/vagrant/ vagrantCentos7#javaPackagingTest
```

I'm pulling and trying again.

nik9000 replied:
Yeah. That command fails for me.

andyb-elastic replied:
Weird, still not seeing it. If you still have it mind sharing the test seed it used

nik9000 replied:
I got this:
```
REPRODUCE WITH: ./gradlew :qa:vagrant:vagrantCentos7#javaPackagingTest -Dtests.seed=924BDF23E0249263 
```

Though I'm not sure how useful that is.

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/util/Cleanup.java:
```diff
@@ -0,0 +1,120 @@
+    private static void purgePackagesLinux() {
+        final Shell sh = new Shell();
+
+        if (isRPM()) {
```
Aren't these all mutually exclusive?

andyb-elastic replied:
They should be (at least deb vs rpm). That said I think the logic here does make sense (if any of these package managers exist on the system, remove the packages from them). It's the same [in the bats tests](https://github.com/elastic/elasticsearch/blob/a8cea90e1084983b2d045edae84187fe8708399d/qa/vagrant/src/test/resources/packaging/utils/utils.bash#L289-L307)

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/util/Shell.java:
```diff
@@ -0,0 +1,181 @@
+        this.workingDirectory = workingDirectory;
+    }
+
+    public Result runAndAssertSuccess(String... command) {
```
I think it is far more often we want to assert success. Can we flip this around have have bare `run` assert, and have a `runIgnoreFailure` for the other case?

andyb-elastic replied:
Yeah good point, it's also probably more intuitive that way

rjernst reviewed (approved):
This looks pretty good. I think it would be good to have some type of documentation on how to write/extend these tests, as there are a number of classes now that one must understand to change these tests. But that can be in followups.

andyb-elastic committed:
rename shell run methods

andyb-elastic committed:
forbiddenapis fixes

The use of java.io.File seems unavoidable because that's the type that
processbuilder expects when setting a working directory, there doesn't
appear to be an alternative that uses Path

We can't know ahead of time what character set subprocesses' output will
use. It seems best to use the jvm's default character set, which is
based on the platform's character set. For all linux platforms we test
that's utf8, and for the windows platforms we test that's iso-8859-1 aka
windows codepage 1252; in all cases this is reflected correctly in the
jvm's default character set

andyb-elastic committed:
Merge branch 'master' into feature-java-archive-tests-1

andyb-elastic commented:
Talked to Nik and he said he's ✅ on this

andyb-elastic commented:
@rjernst sure thing can do a follow up with some docs about writing tests for this project. Where's the right place for those to go? Readme in `qa/vagrant`?

rjernst commented:
> Where's the right place for those to go? Readme in qa/vagrant?

That's a good place to start.
